aries/aries_vcx_core/src/wallet/askar/packing.rs

```diff
@@ -0,0 +1,303 @@
+    wallet::{
+        askar::crypto_box::{CryptoBox, SodiumCryptoBox},
+        structs_io::UnpackMessageOutput,
+    },
+};
+
+pub struct Packing {
+    crypto_box: Box<dyn CryptoBox + Send + Sync>,
+}
+
+impl Packing {
+    pub fn new() -> Self {
+        Self {
+            crypto_box: Box::new(SodiumCryptoBox::new()),
+        }
+    }
+
+    pub async fn unpack(
+        &self,
+        jwe: Jwe,
+        session: &mut Session,
+    ) -> VcxCoreResult<UnpackMessageOutput> {
+        let protected_data = self.unpack_protected_data(&jwe)?;
+        let (recipient, key_entry) = self.find_recipient_key(&protected_data, session).await?;
+        let local_key = key_entry.load_local_key()?;
+        let (enc_key, sender_verkey) = self.unpack_recipient(recipient, &local_key)?;
+        Ok(UnpackMessageOutput {
+            message: self.unpack_msg(&jwe, enc_key)?,
+            recipient_verkey: recipient.unwrap_kid().to_owned(),
+            sender_verkey: sender_verkey.map(|key| key.base58()),
+        })
+    }
+
+    pub fn unpack_protected_data(&self, jwe: &Jwe) -> VcxCoreResult<ProtectedData> {
+        from_json_str(&jwe.protected.decode_to_string()?)
+    }
+
+    pub fn unpack_msg(&self, jwe: &Jwe, enc_key: LocalKey) -> VcxCoreResult<String> {
+        let ciphertext = &jwe.ciphertext.decode()?;
+        let tag = &jwe.tag.decode()?;
+
+        bytes_to_string(
+            enc_key
+                .aead_decrypt(
+                    ToDecrypt::from((ciphertext.as_ref(), tag.as_ref())),
+                    &jwe.iv.decode()?,
+                    &jwe.protected.as_bytes(),
+                )?
+                .to_vec(),
+        )
+    }
+
+    fn unpack_recipient(
+        &self,
+        recipient: &Recipient,
+        local_key: &LocalKey,
+    ) -> VcxCoreResult<(LocalKey, Option<Key>)> {
+        match recipient {
+            Recipient::Authcrypt(auth_recipient) => {
+                self.unpack_authcrypt(local_key, auth_recipient)
+            }
+            Recipient::Anoncrypt(anon_recipient) => {
+                self.unpack_anoncrypt(local_key, anon_recipient)
+            }
+        }
+    }
+
+    fn unpack_authcrypt(
+        &self,
+        local_key: &LocalKey,
+        recipient: &AuthcryptRecipient,
+    ) -> VcxCoreResult<(LocalKey, Option<Key>)> {
+        let (private_bytes, public_bytes) = ed25519_to_x25519_pair(local_key)?;
+        let sender_vk_vec = self.crypto_box.sealedbox_decrypt(
+            &private_bytes,
+            &public_bytes,
+            &recipient.header.sender.decode()?,
+        )?;
+        Ok((
+            LocalKey::from_secret_bytes(
+                KeyAlg::Chacha20(Chacha20Types::C20P),
+                &self.crypto_box.box_decrypt(
+                    &private_bytes,
+                    &ed25519_to_x25519_public(&LocalKey::from_public_bytes(
+                        Ed25519,
+                        &bs58_to_bytes(&bytes_to_string(sender_vk_vec.clone())?)?,
```

Comment:
seem like unnecessary conversion doing `bytes_to_string`. The `bs58_to_bytes` can be changed to accept argument `key: &[u8]` (which is what it needs internally), rather than its current `key: &str`.